Refactor dispatcher and native to use Signature structure.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45990 Refactor dispatcher and native to use Signature structure.**
* #45975 Remove unnecessary byte-for-byte compatibility code that is not needed.
* #45974 Rename legacy_dispatcher to native.
* #45973 Reorder dispatcher/legacy_dispatcher types
* #45918 Add NativeFunctionGroup
* #45890 Rewrite implementation of faithful cpp signatures

In #45890 we introduced the concept of a CppSignature, which bundled
up all of the information necessary to declare a C++ signature for
the cpp API.  This PR introduces analogous concepts for dispatcher
and native: DispatcherSignature and NativeSignature.

The three interfaces are not particularly well coupled right now,
but they do have some duck typing coincidences:

- defn() which renders the C++ definition "bool f(int x)"
- decl() which renders the C++ declaration "bool f(int x = 2)"
- type() which renders the C++ function type "bool(int)"

Maybe at some point we'll introduce a Protocol, or a supertype.
Many other methods (like arguments()) have varying types.  These
signatures also have some helper methods that forward back to real
implementations in the api modules.  Something to think about is
whether or not we should attempt to reduce boilerplate here or
not; I'm not too sure about it yet.

The net effect is we get to reduce the number of variables we
have to explicitly write out in the codegen, since now these are all
bundled together into a signature.  Something extra special happens
in BackendSelect, where we now dynamically select between dispatcher_sig
and native_sig as "how" the backend select is implemented.

A little bit of extra cleanup:
- Some places where we previously advertised Sequence, we now advertise
  a more informative Tuple.
- defn() may take an optional positional parameter overriding the entire
  name, or a kwarg-only prefix parameter to just add a prefix to the
  name.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24223100](https://our.internmc.facebook.com/intern/diff/D24223100)

ezyang committed:
Refactor dispatcher and native to use Signature structure.

In #45890 we introduced the concept of a CppSignature, which bundled
up all of the information necessary to declare a C++ signature for
the cpp API.  This PR introduces analogous concepts for dispatcher
and native: DispatcherSignature and NativeSignature.

The three interfaces are not particularly well coupled right now,
but they do have some duck typing coincidences:

- defn() which renders the C++ definition "bool f(int x)"
- decl() which renders the C++ declaration "bool f(int x = 2)"
- type() which renders the C++ function type "bool(int)"

Maybe at some point we'll introduce a Protocol, or a supertype.
Many other methods (like arguments()) have varying types.  These
signatures also have some helper methods that forward back to real
implementations in the api modules.  Something to think about is
whether or not we should attempt to reduce boilerplate here or
not; I'm not too sure about it yet.

The net effect is we get to reduce the number of variables we
have to explicitly write out in the codegen, since now these are all
bundled together into a signature.  Something extra special happens
in BackendSelect, where we now dynamically select between dispatcher_sig
and native_sig as "how" the backend select is implemented.

A little bit of extra cleanup:
- Some places where we previously advertised Sequence, we now advertise
  a more informative Tuple.
- defn() may take an optional positional parameter overriding the entire
  name, or a kwarg-only prefix parameter to just add a prefix to the
  name.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
There is ZERO codegen diff on this PR

smessmer commented on tools/codegen/api/types.py:
```diff
@@ -240,6 +242,47 @@ class DispatcherArgument:
         return f"{self.type} {self.name}"
 
+@dataclass(frozen=True)
+class DispatcherSignature:
```
Can you add a class comment describing what the dispatcher signature is for and how it differs from NativeSignature / CppSignature ? Side note: maybe rename CppSignature to CppApiSigntature or CppFrontendSignature

smessmer commented on tools/codegen/api/types.py:
```diff
@@ -240,6 +242,47 @@ class DispatcherArgument:
         return f"{self.type} {self.name}"
 
+@dataclass(frozen=True)
+class DispatcherSignature:
```
Can you add a class comment describing what the dispatcher signature is for and how it differs from NativeSignature / CppSignature ? Same for those classes.

Side note: maybe rename CppSignature to ApiSignature or FrontendSignature since they're all kind of Cpp.

ezyang replied:
This explanation would overlap with the top level comments in `tools.codegen.api.FOO` modules, so I'd prefer not to do it here. Maybe a Note reference would be better?

Wondering maybe we can reorganize these modules a bit so that the types aren't splattered about. One possibility is to get rid of types module and move the type definitions to their respective core modules. That will result in cycles when we reference types from the other modules. We could maybe resolve these cycles using the trick in https://mypy.readthedocs.io/en/stable/common_issues.html#import-cycles This looks kind of ugly so I tried not to do it this way, but maybe it is worth it.

> Side note: maybe rename CppSignature to ApiSignature or FrontendSignature since they're all kind of Cpp.

Vetoing api because then we'll get `tools.codegen.api.api` (I guess we could rename the first `api` in this list, though I'm not sure what too; `cconv` as in calling convention?) Frontend might be OK; there's some potential for confusion with Python frontend but I guess it's not too bad.

ezyang committed:
Update on "Refactor dispatcher and native to use Signature structure."


In #45890 we introduced the concept of a CppSignature, which bundled
up all of the information necessary to declare a C++ signature for
the cpp API.  This PR introduces analogous concepts for dispatcher
and native: DispatcherSignature and NativeSignature.

The three interfaces are not particularly well coupled right now,
but they do have some duck typing coincidences:

- defn() which renders the C++ definition "bool f(int x)"
- decl() which renders the C++ declaration "bool f(int x = 2)"
- type() which renders the C++ function type "bool(int)"

Maybe at some point we'll introduce a Protocol, or a supertype.
Many other methods (like arguments()) have varying types.  These
signatures also have some helper methods that forward back to real
implementations in the api modules.  Something to think about is
whether or not we should attempt to reduce boilerplate here or
not; I'm not too sure about it yet.

The net effect is we get to reduce the number of variables we
have to explicitly write out in the codegen, since now these are all
bundled together into a signature.  Something extra special happens
in BackendSelect, where we now dynamically select between dispatcher_sig
and native_sig as "how" the backend select is implemented.

A little bit of extra cleanup:
- Some places where we previously advertised Sequence, we now advertise
  a more informative Tuple.
- defn() may take an optional positional parameter overriding the entire
  name, or a kwarg-only prefix parameter to just add a prefix to the
  name.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatcher and native to use Signature structure."


In #45890 we introduced the concept of a CppSignature, which bundled
up all of the information necessary to declare a C++ signature for
the cpp API.  This PR introduces analogous concepts for dispatcher
and native: DispatcherSignature and NativeSignature.

The three interfaces are not particularly well coupled right now,
but they do have some duck typing coincidences:

- defn() which renders the C++ definition "bool f(int x)"
- decl() which renders the C++ declaration "bool f(int x = 2)"
- type() which renders the C++ function type "bool(int)"

Maybe at some point we'll introduce a Protocol, or a supertype.
Many other methods (like arguments()) have varying types.  These
signatures also have some helper methods that forward back to real
implementations in the api modules.  Something to think about is
whether or not we should attempt to reduce boilerplate here or
not; I'm not too sure about it yet.

The net effect is we get to reduce the number of variables we
have to explicitly write out in the codegen, since now these are all
bundled together into a signature.  Something extra special happens
in BackendSelect, where we now dynamically select between dispatcher_sig
and native_sig as "how" the backend select is implemented.

A little bit of extra cleanup:
- Some places where we previously advertised Sequence, we now advertise
  a more informative Tuple.
- defn() may take an optional positional parameter overriding the entire
  name, or a kwarg-only prefix parameter to just add a prefix to the
  name.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24223100](https://our.internmc.facebook.com/intern/diff/D24223100)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatcher and native to use Signature structure."


In #45890 we introduced the concept of a CppSignature, which bundled
up all of the information necessary to declare a C++ signature for
the cpp API.  This PR introduces analogous concepts for dispatcher
and native: DispatcherSignature and NativeSignature.

The three interfaces are not particularly well coupled right now,
but they do have some duck typing coincidences:

- defn() which renders the C++ definition "bool f(int x)"
- decl() which renders the C++ declaration "bool f(int x = 2)"
- type() which renders the C++ function type "bool(int)"

Maybe at some point we'll introduce a Protocol, or a supertype.
Many other methods (like arguments()) have varying types.  These
signatures also have some helper methods that forward back to real
implementations in the api modules.  Something to think about is
whether or not we should attempt to reduce boilerplate here or
not; I'm not too sure about it yet.

The net effect is we get to reduce the number of variables we
have to explicitly write out in the codegen, since now these are all
bundled together into a signature.  Something extra special happens
in BackendSelect, where we now dynamically select between dispatcher_sig
and native_sig as "how" the backend select is implemented.

A little bit of extra cleanup:
- Some places where we previously advertised Sequence, we now advertise
  a more informative Tuple.
- defn() may take an optional positional parameter overriding the entire
  name, or a kwarg-only prefix parameter to just add a prefix to the
  name.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24223100](https://our.internmc.facebook.com/intern/diff/D24223100)

[ghstack-poisoned]
